src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,84 @@
+"""
+Automatically scale MDSs based on status of the file-system using the FSMap
+"""
+
+import errno
+import json
+import time
+import mgr_util
+import threading
+import uuid
+import logging
+from six import itervalues, iteritems
+from collections import defaultdict
+from prettytable import PrettyTable, PLAIN_COLUMNS
+from mgr_module import MgrModule
+
+import orchestrator
+
+log = logging.getLogger(__name__)
+
+"""
+Some terminology is made up for the purposes of this module:
+
+ - "grow/shrink" - increase/decrease the MDSs in the system
+"""
+
+class MdsAutoscaler(MgrModule):
+    """
+    MDS autoscaler.
+    """
+    NATIVE_OPTIONS = [
+        'max_mds'
+    ]
+
+    MODULE_OPTIONS = [
+        {
+            'name': 'sleep_interval',
+            'default': str(60),
+        },
+    ]
+
+    def __init__(self, *args, **kwargs):
+        super(MdsAutoscaler, self).__init__(*args, **kwargs)
+        self.fs_map = None
+
+    def create_mds(self, fs_name):
+        spec = orchestrator.StatelessServiceSpec(fs_name)
+        try:
+            completion = self.add_mds(spec)
+            self._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+        except Exception as e:
+            # Don't let detailed orchestrator exceptions (python backtraces)
+            # bubble out to the user
+            self.log.exception("Failed to create MDS daemons")
+            return -errno.EINVAL, "", str(e)
+        return 0, "", ""
+
+    def is_fewer_than_max_mds(self, fs):
+        if len(fs['in']) < fs['max_mds']:
+            return True
+        return False
```

Comment:
It's not that simple. You need to review:

https://github.com/ceph/ceph/blob/master/src/mds/MDSMap.cc#L1025-L1033

and

https://github.com/ceph/ceph/blob/057391854c01c10294c8fe56f4b79973e02b7f17/src/mon/MDSMonitor.cc#L1996

In general, you only really need to ensure that sufficient standbys exist for a file system. The monitors will use standbys to stabilize the file system if possible. You will see an "INSUFFICIENT_STANDBY" warning generated by the monitors if there are not enough standbys for a file system. See also:

https://github.com/ceph/ceph/blob/057391854c01c10294c8fe56f4b79973e02b7f17/src/mds/MDSMap.h#L242-L250

This module should be able to monitor for that via the below `notify` method for the "clog" events. However, this module should just replicate the logic that detects insufficient standbys and not try to parse the clog event message.